### ISSUE TITLE
Numerical instability in log_softmax

### ISSUE THREAD

**Issue by aosokin** [pytorch/pytorch#1546]
Hi, it looks like the current GPU implementation of log_softmax is not very numerically stable.
Code
```
import torch
from torch.autograd import Variable
input = torch.cuda.FloatTensor([0.1232, 0.0869, -0.0047, 0.1211, 0.0513, 0.0545,
                                0.0151, -0.0354, 0.0453, -0.0031, 0.0412, -0.0250,
                                -0.0522, -0.0889, -0.0904, -0.0834, 0.0259, -0.0785,
                                0.0016, 0.0152, 0.1067, -0.0039, 0.0144, -0.0634,
                                0.0839, 0.0226, 0.0547,
                                ]).unsqueeze(0)
def compute_log_softmax(input):
    max_vals, max_pos = torch.max(input, 1)
    input = input - max_vals.expand_as(input)
    input_exp = torch.exp(input)
    norm_vals = input_exp.sum(1)
    norm_vals = torch.log(norm_vals)
    # subtract sum
    input = input - norm_vals.expand_as(input)
    return input
builtin = torch.nn.functional.log_softmax(Variable(input))
mine = compute_log_softmax(input)
print(builtin.data - mine)
```
produces 
```
Columns 0 to 9 
1.00000e-07 *
  0.0000  0.0000  0.0000  0.0000  0.0000  2.3842  0.0000  2.3842  0.0000  2.3842

Columns 10 to 19 
1.00000e-07 *
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000

Columns 20 to 26 
1.00000e-07 *
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
[torch.cuda.FloatTensor of size 1x27 (GPU 0)]
```
In my application, this discrepancy costed me 1% of final accuracy.
Is it possible to implement the max trick for stability on low level?

**Comment by apaszke** [comment]
The implementation is already using the max trick in all LogSoftmax implementations (see e.g. [here](https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/LogSoftMax.cu#L258-L269)). The differences seem small (max 2.38e-7) and may be a result of performing the operations in a different order. From what I see the CUDA kernel does `input - (max + log(exp_sum))` while you do `input - max - log(exp_sum)`.

**Comment by aosokin** [comment]
@apaszke Thanks, sorry for bugging.
